src/core/public/deprecations/deprecations_service.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { CoreService } from 'src/core/types';
+import { CoreSetup } from 'src/core/public';
+import { DeprecationsClient } from './deprecations_client';
+
+/**
+ * DeprecationsService provides methods to fetch plugin deprecation details from
+ * the Kibana server.
+ *
+ * @public
+ */
+export interface DeprecationsServiceSetup {
+  /**
+   * Grabs deprecations for all plugins.
+   *
+   * @param {Object} configs { skipCache: boolean };
+   * set skipCache: true to fetch a fresh copy of the deprecations from the kibana server.
+   */
+  getAllDeprecations: DeprecationsClient['getAllDeprecations'];
+  /**
+   * Grabs deprecations for a specific plugin.
+   *
+   * @param {string} pluginId
+   * @param {Object} configs { skipCache: boolean };
+   * set skipCache: true to fetch a fresh copy of the deprecations from the kibana server.
+   */
+  getDeprecations: DeprecationsClient['getDeprecations'];
+}
+
+/**
+ * DeprecationsService provides methods to fetch plugin deprecation details from
+ * the Kibana server.
+ *
+ * @public
+ */
+export type DeprecationsServiceStart = DeprecationsServiceSetup;
+
+export class DeprecationsService
+  implements CoreService<DeprecationsServiceSetup, DeprecationsServiceStart> {
+  private deprecationsClient?: DeprecationsClient;
+
+  public setup({ http }: Pick<CoreSetup, 'http'>): DeprecationsServiceSetup {
+    this.deprecationsClient = new DeprecationsClient({ http });
+    return {
+      getAllDeprecations: this.deprecationsClient.getAllDeprecations,
+      getDeprecations: this.deprecationsClient.getDeprecations,
+    };
```

Comment:
Good catch! I was not sure so I exposed it to both.

Reply:
Ah no worries, this is probably pretty tribal knowledge.

Two reasons we've been doing this:
1. Keep the types of logic that live in each lifecycle hook predictable
2. Optimizing for fast Kibana startup time

#### Keeping lifecycle hooks predictable

We've tried to force "setup" to be when things are initialized and "start" to be when things start executing. Since I don't imagine these APIs will be used when configuring things, then we can prevent developers from using this API until "start". [RFC-0001](https://github.com/elastic/kibana/blob/master/rfcs/text/0001_lifecycle_setup.md) talks more about this.

#### Optimizing for startup times
We've been trying to optimize for fast startup times for operational reasons since it allows Kibana to start reporting its status sooner to users and external systems. This allows us to also build UI tools for repairing Kibana when things go wrong. [RFC-0007](https://github.com/elastic/kibana/blob/master/rfcs/text/0007_lifecycle_unblocked.md) touches on some of the origins of this idea.